### PR TITLE
Fix #188 and more things found during testing

### LINKED ISSUE #188: Sanitize *every* data that user send to us
Like https://jirapdi.tid.es/browse/OWD-23869 or https://jirapdi.tid.es/browse/OWD-23853

### PR DESCRIPTION
**Pull request by willyaranda**

### CODE REVIEW

File: src/common/datastore.js [hunk 1]
```diff
@@ -40,7 +40,10 @@ var DataStore = function() {
         {
           rs_name: ddbbsettings.replicasetName,
           read_secondary: true,
-          w: 1
+          w: 1,
+          socketOptions: {
+            keepalive: 200
```
Comment (willyaranda) [inline src/common/datastore.js]: keepalive for MongoDB

File: src/ns_as/apis/SimplePushAPI_v1.js [hunk 2]
```diff
@@ -90,6 +92,13 @@ var SimplePushAPI_v1 = function() {
     //Now, we are safe to start using the path and data
     log.notify('appToken=' + appToken + ' -- version=' + version);
     dataStore.getChannelIDForAppToken(appToken, function(error, channelID) {
+      // If there is no channelID associated with a appToken,
+      // fool the sender with a OK response, but nothing is done here.
+      if (!channelID) {
+        response.statusCode = 200;
+        response.end('{}');
+        return;
+      }
```
Comment (willyaranda) [inline src/ns_as/apis/SimplePushAPI_v1.js:101]: I talked with Antonio and agreed on this.

Case: a flood attack to know what URIs are registered. If we sent a 404, that means that nobody is there, but sending 200 to all, flooders don't know if it's correct or not.

File: src/ns_ua/ws_server.js [hunk 3]
```diff
@@ -367,7 +366,7 @@ server.prototype = {
                 setTimeout(function recoveryChannels() {
                   log.debug('WS::onWSMessage::recoveryChannels --> Recovery channels process: ', query);
                   // TODO sync channels with client
-                });
+                }, 300); //Enough time to register and replicate data
```
Comment (willyaranda) [inline src/ns_ua/ws_server.js]: 300ms to wait for replication is not a long time. And should not interfere with the service.

File: src/ns_ua/ws_server.js [hunk 4]
```diff
@@ -375,15 +374,21 @@ server.prototype = {
                 log.debug('WS::onWSMessage::pendingNotifications --> Sending pending notifications');
                 dataManager.getNodeData(query.uaid, function(err, data) {
                   if (err) {
+                    log.error('WS::onWSMessage::pendingNotifications --> There was an error getting the node');
+                    return;
+                  }
+                  // In this case, there are no nodes for this (strange, since it was just registered)
+                  if (!data || !data.ch || !Array.isArray(data.ch)) {
+                    log.error('WS::onWSMessage::pendingNotifications --> No channels for this node.');
```
Comment (willyaranda) [inline src/ns_ua/ws_server.js]: This fixes what @carlosmartineztoral reports

File: src/ns_ua/connectors/connector.js [hunk 5]
```diff
@@ -21,14 +21,21 @@ Connector.prototype = {
    * Create and return a connector object based on the data received
    */
   getConnector: function(data, connection, callback) {
+    //Is valid IP?
+    var ip = data.wakeup_hostport && data.wakeup_hostport.ip &&
```
Comment (willyaranda) [inline src/ns_ua/connectors/connector.js:25]: This will return `0` if it's not a IP, `4` if it's IPv4 and `6` if it is IPv6.

**Comment by frsela** [comment]
Rebase, please.

**Comment by willyaranda** [comment]
Merged with rebase. https://github.com/telefonicaid/notification_server/commit/05b525541fd8257e06e3f63ec6502fe39d074160

**Pull request closed without merging**